pkg/util/updatesort/scatter_sort.go

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2019 The Kruise Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package updatesort
+
+import (
+	"sort"
+
+	appsv1alpha1 "github.com/openkruise/kruise/pkg/apis/apps/v1alpha1"
+	v1 "k8s.io/api/core/v1"
+)
+
+type scatterSort struct {
+	strategy appsv1alpha1.CloneSetUpdateScatterStrategy
+}
+
+func NewScatterSorter(s appsv1alpha1.CloneSetUpdateScatterStrategy) Sorter {
+	return &scatterSort{strategy: s}
+}
+
+// Sort helps scatter the indexes of pods by ScatterStrategy.
+func (ss *scatterSort) Sort(pods []*v1.Pod, indexes []int) []int {
+	if len(ss.strategy) == 0 {
+		return indexes
+	}
+
+	terms := ss.generateRules(pods, indexes)
+	for _, term := range terms {
+		indexes = ss.scatterPodsByRule(term, pods, indexes)
+	}
+	return indexes
+}
+
+func (ss *scatterSort) generateRules(pods []*v1.Pod, indexes []int) []appsv1alpha1.CloneSetUpdateScatterTerm {
+	var termSlice []appsv1alpha1.CloneSetUpdateScatterTerm
+	ruleCounter := map[string]int{}
+
+	termID := func(term appsv1alpha1.CloneSetUpdateScatterTerm) string {
+		return term.Key + ":" + term.Value
+	}
+
+	for _, term := range ss.strategy {
+		for _, idx := range indexes {
+			if val, ok := pods[idx].Labels[term.Key]; ok {
+				if term.Value == "*" || val == term.Value {
+					term := appsv1alpha1.CloneSetUpdateScatterTerm{Key: term.Key, Value: val}
```

Comment:
Nit: you may replace temp variable term with ss.strategy[i].

Reply:
There is no need to do so, but i will change the temp variable name for more readable.